### PR TITLE
Rebuild the Drupal route cache.

### PR DESCRIPTION
**Pull request by homotechsual**
Overview
----------------------------------------
_This is linked to [issue 158](https://lab.civicrm.org/dev/drupal/-/issues/158) and [issue 37](https://lab.civicrm.org/dev/drupal/-/issues/37)._

_It takes an alternative approach to #19888, adding a Utils/System function to rebuild the routes which we can then invoke from multiple places if required with an invocation added to the rebuildMenuandCaches function in invoke.php._

Before
----------------------------------------
_Enabling a CiviCRM extension that adds a route on Drupal 8 or 9 results in a 404 until the routes are rebuilt in Drupal._

After
----------------------------------------
_Enabling/disabling an extension rebuilds Drupal's routes and routes added in CiviCRM work as expected.._

Technical Details
----------------------------------------
_There might be a more performant option for this but I can't find anything..._

### CODE REVIEW

**Comment by MikeyMJCO** [comment]
Alternative approach for #19888

**Comment by MikeyMJCO** [comment]
Test sites:

D8: https://d8pr19906.bkonaz.mjco.uk/
D9: https://d9pr19906.bkonaz.mjco.uk/

**Comment by demeritcowboy** [comment]
I think what @seamuslee001  meant on the other PR by CRM_Utils_System_x classes is that instead of hardcoding Drupal8, here in Invoke you would call `$config->userSystem->invalidateRouteCache()` and in CRM_Utils_System_Base you would put a do-nothing `function invalidateRouteCache() { return; }` and in CRM_Utils_System_Drupal8 you would override that function.

**Comment by MikeyMJCO** [comment]
> I think what @seamuslee001 meant on the other PR by CRM_Utils_System_x classes is that instead of hardcoding Drupal8, here in Invoke you would call `$config->userSystem->invalidateRouteCache()` and in CRM_Utils_System_Base you would put a do-nothing `function invalidateRouteCache() { }` and in CRM_Utils_System_Drupal8 you would override that function.

That makes sense.

**Comment by MikeyMJCO** [comment]
Refactored with @demeritcowboy 's suggested route and squashed.

File: CRM/Core/Invoke.php [hunk 1]
```diff
@@ -394,6 +394,7 @@ public static function rebuildMenuAndCaches($triggerRebuild = FALSE, $sessionRes
       CRM_Utils_Request::retrieve('triggerRebuild', 'Boolean', CRM_Core_DAO::$_nullObject, FALSE, 0, 'GET')
     ) {
       CRM_Core_DAO::triggerRebuild();
+      $config->userFramework->invalidateRouteCache();
```
Comment (demeritcowboy) [inline CRM/Core/Invoke.php:397]: Typo I think. Should be userSystem. `userFramework` is a string.

**Comment by demeritcowboy** [comment]
I tested this in drupal 9 with cdntaxreceipts but I still needed to clear drupal cache after in order to avoid the "page not found". Hmmmm.

**Comment by MikeyMJCO** [comment]
> I tested this in drupal 9 with cdntaxreceipts but I still needed to clear drupal cache after in order to avoid the "page not found". Hmmmm.

If you switch it out to do `\Drupal::service('router.builder')->rebuild();` do you have to clear the Drupal Cache?

**Comment by MikeyMJCO** [comment]
Okay this works now, it turns out we do need to do a `router.builder` rebuild for this to work!

**Comment by demeritcowboy** [comment]
Ok neat. I guess we're then back to where is the best place. So this would trigger on any extension action even disable but maybe that's ok/desired.

**Comment by MikeyMJCO** [comment]
Yeah, I'm thinking that having extension routes in the cache post-disable is undesirable so it's cleaner to run this on any extension enable/disable/install/uninstall.

**Comment by demeritcowboy** [comment]
* General standards
    * [`r-explain`] __PASS__
    * [`r-user`] __PASS__
    * [`r-doc`] __PASS__
    * [`r-run`] __PASS__
* Developer standards
    * [`r-tech`] __PASS__
      * It seems like this would only be called during extension lifecycle events or some already heavy cache clearing event. I can't think of a better place to put it unless it's left to extensions to decide when to do it.
    * [`r-code`] __PASS__
      * I don't know that it needs to call class_exists('\Drupal') given that this is already within a drupal8-only file. Ditto hasContainer() although I suppose that call might be a safeguard against some weird edge case.
    * [`r-maint`] __?__
    * [`r-test`] __PASS__

**Comment by MikeyMJCO** [comment]
I went with checking for `\Drupal` and `->hasContainer()` because the cost is minimal of those checks and it would catch any possible scenario where this tries to get invoked during installs or in situations where Drupal isn't booted yet.

**Comment by eileenmcnaughton** [comment]
Thanks for the review @demeritcowboy - I think the r-maint is OK since it's just a little bit of code that seems pretty readable

**Comment by eileenmcnaughton** [comment]
@MikeyMJCO this is 'linked to issues' - are they closable now or just linked

**Comment by MikeyMJCO** [comment]
Closeable along with PR #19888

**Comment by MikeyMJCO** [comment]
> @MikeyMJCO this is 'linked to issues' - are they closable now or just linked

Issues and other PR closed :-)

**Comment by eileenmcnaughton** [comment]
thanks

**Comment by eileenmcnaughton** [comment]
@MikeyMJCO - although I see I said above the code is pretty readable I'm reading it now & I'm confused :-)

Specifically I don't understand why it's nested in the if trigger part here - because it seems the line is about rebuilding paths which is 'all the rest of the function' rather than re-instating logging

image


If it IS correct for it to be in that loop maybe you could add a PR to explain why it should only be done when rebuilding mysql triggers

**Comment by MikeyMJCO** [comment]
It's a computationally very expensive operation - like rebuilding triggers. It's not something that should be called as often as we flush caches - the performance cost would be very high.

When we do things like disable/enable or upgrade extensions we're already setting `triggerRebuild` to `TRUE` so it felt more appropriate to put it in that conditional than call it any time we rebuild the cache.

https://github.com/civicrm/civicrm-core/blob/c706f46856429fc31af88b4b712860045aaf7496/CRM/Extension/Manager.php#L228

**Comment by MikeyMJCO** [comment]
Described in comments in #20414
